src/sentry/snuba/metrics/query_builder.py

```diff
@@ -251,13 +317,30 @@ def resolve_tags(
         if is_tag_value:
             return resolve_tag_value(use_case_id, org_id, input_)
         else:
+            expression, is_restricted = is_tag_key_restricted(input_, tag_keys_restrictions)
+            if is_restricted:
+                raise InvalidParams(
+                    f"The tag key ${input_} usage has been restricted by the ${expression} expression"
+                )
+
             return resolve_weak(use_case_id, org_id, input_)
     if isinstance(input_, int):
         return input_
 
     raise InvalidParams("Unable to resolve conditions")
 
 
+def is_tag_key_restricted(
+    tag_key: str, tag_keys_restrictions: Optional[Dict[str, str]]
+) -> Tuple[Optional[str], bool]:
+    if tag_keys_restrictions is None:
+        return None, False
+
+    expression = tag_keys_restrictions.get(f"tags[${tag_key}]")
+
+    return expression, expression is not None
```

Comment:
Seems slightly over complicated ( don't have super strong opinion but I would implemented like this):

Also I think the `$` is from typescript interpolation.

```
def is_tag_key_restricted(tag_key: str, tag_keys_restrictions: Optional[Dict[str, str]])->Optional[str]:
    """
    >>> is_tag_key_restricted("f1", restr)
    'f1'
    >>> is_tag_key_restricted("f2", restr)
    'f2'
    >>> is_tag_key_restricted("f3", restr) == None
    True
    >>> is_tag_key_restricted("f1", None) == None
    True
    """
    if tag_keys_restrictions is None:
        return None
    return tag_keys_restrictions.get(f"tags[{tag_key}]")

```

Reply:
You are right, I overlooked the redundancy, in the next PR I will have a slightly different logic nontheless.